### PR TITLE
orchestrator-process: export a prometheus service discovery configuration

### PR DESCRIPTION
**Pull request by benesch**
The tl;dr is that this PR makes it possible to run Prometheus locally and have it auto scrape all clusters for metrics. See commits for details.

### Motivation

  * This PR adds a development feature that was discussed on Slack.

### Tips for reviewer

Suppress whitespace. Had to indent a bunch of stuff to make the borrow checker happy with `MutexGuard` lifetimes.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

File: misc/prometheus/mzcompose.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+from materialize.mzcompose.services import Service
+
+SERVICES = [
+    Service(
```
Comment (philip-stoev) [inline misc/prometheus/mzcompose.py:13]: Should this service definition be put in mzcompose/services.py ? I guess the end goal would be to not have to mess with a given mzcompose workflow to get prometheus up and running. The way we currently have `--redpanda` for several workflows, we should also eventually make it possible to just have a `--prometheus` option to any workflow that will do the required stuff for you.

File: misc/prometheus/mzcompose.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        "prometheus",
+        {
+            "image": "prom/prometheus:v2.41.0",
+            "ports": ["9090:9090"],
+            "volumes": [
+                "./prometheus.yml:/etc/prometheus/prometheus.yml",
+                "../../mzdata/prometheus:/mnt/services",
```
Comment (philip-stoev) [inline misc/prometheus/mzcompose.py:20]: I am afraid there is a permission issue of some sort. The mzdata directory is created, owned by `root`

```
pstoev@Ubuntu-2004-focal-64-minimal:~/benesch-process-monitoring$ ls -la mzdata/prometheus/
total 8
drwxr-xr-x 2 root root 4096 Dec 27 10:34 .
drwxr-xr-x 3 root root 4096 Dec 27 10:34 ..
```

And therefore bin/environmentd is unable to start:
```
Traceback (most recent call last):
  File "/home/pstoev/benesch-process-monitoring/misc/python/materialize/cli/run.py", line 137, in main
    environment_id = environment_file.read_text().rstrip()
  File "/usr/lib/python3.9/pathlib.py", line 1266, in read_text
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
  File "/usr/lib/python3.9/pathlib.py", line 1252, in open
    return io.open(self, mode, buffering, encoding, errors, newline,
  File "/usr/lib/python3.9/pathlib.py", line 1120, in _opener
    return self._accessor.open(self, flags, mode)
FileNotFoundError: [Errno 2] No such file or directory: '/home/pstoev/benesch-process-monitoring/mzdata/environment-id'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/pstoev/benesch-process-monitoring/misc/python/materialize/cli/run.py", line 239, in <module>
    main()
  File "/home/pstoev/benesch-process-monitoring/misc/python/materialize/cli/run.py", line 140, in main
    environment_file.write_text(environment_id)
  File "/usr/lib/python3.9/pathlib.py", line 1285, in write_text
    with self.open(mode='w', encoding=encoding, errors=errors) as f:
  File "/usr/lib/python3.9/pathlib.py", line 1252, in open
    return io.open(self, mode, buffering, encoding, errors, newline,
  File "/usr/lib/python3.9/pathlib.py", line 1120, in _opener
    return self._accessor.open(self, flags, mode)
PermissionError: [Errno 13] Permission denied: '/home/pstoev/benesch-process-monitoring/mzdata/environment-id'
```

**Review by philip-stoev — APPROVED** [review]

File: misc/prometheus/mzcompose.py (hunk 2 above)
Comment (benesch) [reply misc/prometheus/mzcompose.py:20]: Ah, yeah, you have to start them in the opposite order: first `bin/environmentd`, and _then_ `./mzcompose run default`. I think we can fix that by creating the directory before the Docker daemon though, in mzcompose. Sec.

**Review by danhhz — APPROVED** [review]
amazing! \o/

I'm a bit 🔬🐶 on bits of this, but scanned it and everything seems fine. Tried it locally and verified that it works. Also wrote up a quick set of instructions for stealing the persist prod dash and putting it in a local grafana pointed at this.

File: misc/python/materialize/cli/run.py [hunk 3]
```diff
@@ -143,6 +143,7 @@ def main() -> int:
                 f"--orchestrator=process",
                 f"--orchestrator-process-secrets-directory={mzdata}/secrets",
                 f"--orchestrator-process-tcp-proxy-listen-addr=127.0.0.1",
+                f"--orchestrator-process-prometheus-service-discovery-directory={mzdata}/prometheus",
```
[the remark this answers is not in the document]
Comment (benesch) [reply misc/python/materialize/cli/run.py:146]: Ah, seems easy enough to stick `args.args` at the end instead of the beginning. Done.

File: src/orchestrator-process/src/lib.rs [hunk 4]
```diff
@@ -520,6 +564,51 @@ impl NamespacedProcessOrchestrator {
...
+        let Some(dir) = &tcp_proxy.prometheus_service_discovery_dir else { return };
+
+        let mut static_configs = vec![];
+        {
+            let services = self.services.lock().expect("lock poisoned");
+            for (id, states) in &*services {
+                for (i, state) in states.iter().enumerate() {
+                    for (name, port) in &state.tcp_proxy_ports {
+                        let mut labels = btreemap! {
+                            "namespace".into() => self.namespace.clone(),
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/orchestrator-process/src/lib.rs:585]: Ah, good call. Adjusted to prefix all of these labels with "mz_orchestrator".

File: misc/prometheus/mzcompose.py (hunk 2 above)
Comment (benesch) [reply misc/prometheus/mzcompose.py:22]: Fixed!

File: src/environmentd/src/bin/environmentd/main.rs [hunk 5]
```diff
@@ -387,6 +389,23 @@ pub struct Args {
...
     orchestrator_process_tcp_proxy_listen_addr: Option<IpAddr>,
+    /// A directory in which the process orchestrator should write Prometheus
+    /// scrape targets, for use with Prometheus's file-based service discovery.
+    ///
+    /// Each namespaced orchestrator will maintain a single JSON file into the
+    /// directory named `NAMESPACE.json` containing the scrape targets for all
+    /// extant services. The scrape targets will use the TCP proxy address, as
+    /// Prometheus does not support scraping over Unix domain sockets.
+    ///
+    /// This option is ignoed unless `--orchestrator-tcp-proxy-listen-addr` is
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/environmentd/src/bin/environmentd/main.rs]: Thanks, fixed the typo. Agree the silent ignoring is bad, but there's no built in support for expressing this constraint in Clap that I can find. There's a few other arguments at the moment that work similarly, so just going to leave as is for now, and do a pass at some point in the future to clean them all up uniformly.

File: src/orchestrator-process/src/lib.rs [hunk 6]
```diff
@@ -520,6 +564,51 @@ impl NamespacedProcessOrchestrator {
...
+                            "service_id".into() => id.clone(),
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/orchestrator-process/src/lib.rs]: I think "pod" is the concatenation of namespace and service ID. I don't want to pollute the process orchestrator with Kubernetes concepts, but I added a relabeling to the Prometheus service config!

[the remark this answers is not in the document]
Comment (danhhz) [reply src/orchestrator-process/src/lib.rs]: wfm!

**Comment by benesch** [comment]
Pushed up a modification to run Grafana too, based on your playbook! 🎉
